packages/hardhat/contracts/DonationMultiSig.sol

```diff
@@ -0,0 +1,150 @@
+	}
+	mapping(address => Candidate) proposals;
+
+	address[] public contributors;
+
+	uint32[] public weights;
+	uint256 public totalWeight;
+
+	event Donation(address sender, uint256 amount);
+
+	modifier onlyContributors() {
+		bool contains;
+		for (uint8 i = 0; i < contributors.length; i++) {
+			if (msg.sender == contributors[i]) {
+				contains = true;
+				break;
+			}
+		}
+		require(contains, "Only contributors can make this call");
+		_;
+	}
+
+	constructor(address[] memory _contributors, uint32[] memory _weights) {
+		require(
+			_contributors.length == _weights.length,
+			"Array length mismatch"
+		);
+		for (uint8 i = 0; i < _contributors.length; i++) {
+			require(_weights[i] != 0, "Zero weight not allowed");
+			contributors.push(_contributors[i]);
+			weights.push(_weights[i]);
+			totalWeight += _weights[i];
+		}
+	}
+
+	fallback() external payable {
+		_donate();
+	}
+
+	receive() external payable {
+		_donate();
+	}
+
+	function proposeContributor(
+		address newContributor,
+		uint32 weight
+	) external onlyContributors {
+		if (weight == 0) {
+			revert NewContributorCannotHaveZeroWeight(newContributor);
+		}
+
+		if (!proposals[newContributor].exists) {
+			if (_isContract(newContributor)) {
+				revert NewContributorCannotBeContract(newContributor);
+			}
+			proposals[newContributor].exists = true;
+		} else {
+			if (weight == proposals[newContributor].weight) {
+				revert NewContributorAlreadyProposed(newContributor);
+			}
+
+			// If proposed weight is different, then reset all approvals
+			for (uint8 i = 0; i < contributors.length; i++) {
+				proposals[newContributor].approvals[contributors[i]] = false;
+			}
+		}
+		proposals[newContributor].approvals[msg.sender] = true;
+		proposals[newContributor].weight = weight;
+	}
+
+	function approveContributor(
+		address newContributor
+	) external onlyContributors {
+		if (!proposals[newContributor].exists) {
+			revert ProposalDoesntExist(newContributor);
+		}
+		if (proposals[newContributor].approvals[msg.sender]) {
+			revert NewContributorAlreadyApproved(newContributor, msg.sender);
+		}
+		proposals[newContributor].approvals[msg.sender] = true;
+	}
+
+	function _addContributor(address newContributor) internal {
+		for (uint8 i = 0; i < contributors.length; i++) {
+			if (!proposals[newContributor].approvals[contributors[i]]) {
+				return;
+			}
+		}
+		contributors.push(newContributor);
+		weights.push(proposals[newContributor].weight);
+		totalWeight += proposals[newContributor].weight;
+
+		// Remove the proposal
+		for (uint8 i = 0; i < contributors.length; i++) {
+			proposals[newContributor].approvals[contributors[i]] = false;
+		}
+		proposals[newContributor].exists = false;
+		proposals[newContributor].weight = 0;
+	}
+
+	function _donate() internal {
+		if (msg.value == 0) {
+			return;
+		}
+
+		uint256 totalWeiToSend = address(this).balance;
+		if (totalWeiToSend < totalWeight) {
+			// Not enough balance, not worth splitting
+			return;
+		}
```

Comment:
Just wanted to say that the above five lines are so simply genius. Nice work. There will be minuscule amounts left in the contract but it is not a problem because the contract is guaranteed to send what it can without encountering division issues. I will be duplicating this kind of simple yet effective solution in the future.